### DISCUSSION TITLE
scrollbar layout shift with dialog, dropdownmenu, etc

### DISCUSSION THREAD

**Discussion by rtkaaho** [radix-ui/primitives#1100] [Help] [answered]
For components that remove pointer events when opened, I am noticing a decent amount of layout shift as a result of the scrollbars toggling in and out. That is not something I am noticing on the radix docs site. I was wondering if you could provide any tips/tricks that you have learned to remove some of that layout shift?

**Comment by benoitgrelard** [comment, marked as answer]
Hey @rtkaaho,

It is actually due to removing the scroll, rather than pointer events.
Under the hood, we use a library called `react-remove-scroll` which exposes some classes to use to mitigate this.

If you look at our docs code, you'll see [some examples](https://github.com/radix-ui/website/search?q=RemoveScroll) of how we use this in places.

Worth looking at also, the [lib itself](https://github.com/theKashey/react-remove-scroll/blob/master/src/UI.tsx#L98-L101).
It is itself built on top of a more generic one, where you can see [the constants exported](https://github.com/theKashey/react-remove-scroll-bar/blob/master/src/constants.ts).

I hope that helps!

**Reply by rtkaaho** [reply]
Thanks!!

**Reply by jjenzz** [reply]
Aw man, I can't wait till we can provide a nicer API for all this 🙈

**Reply by hamidbjss** [reply]
@benoitgrelard I'm noticing this issue in Windows even when my component (either Dropdown or Select) is not using fixed positioning. 

I've tried using the `react-remove-scroll` lib as you've mentioned in the docs to no avail (although it does say it's to address fixed positioning errors). 

Any other solutions to resolve the issue with page layout shift in Windows when opening/closing Select and Dropdown components?

**Reply by benoitgrelard** [reply]
@hamidbjss the issue you're seeing might actually be due to a regression in this lib that we are trying to get resolved.
See here: https://github.com/theKashey/react-remove-scroll-bar/pull/32

**Reply by hamidbjss** [reply]
@benoitgrelard yep, that seems to be describing what I'm facing. Thanks for getting back to me, will keep an eye on it!

**Reply by mithleshjs** [reply]
@hamidbjss Try this solution https://github.com/radix-ui/primitives/issues/346#issuecomment-1121440376

**Reply by DonHuskini** [reply]
The solution that worked for me was:

```css
body {
  margin: 0;
}
```
```jsx
function onOpenChange(open: boolean) {
    document.body.style.paddingRight = open ? "0px" : "";
}
return (
  <DropdownMenu.Root onOpenChange={onOpenChange}>
    {children}
  </DropdownMenu.Root>
);

```

**Reply by benoitgrelard** [reply]
FYI, we have fixed the issue in the upstream package and upgraded the dependency. So the regression should be gone in all the latest release candidates.

https://github.com/radix-ui/primitives/pull/1448

**Reply by hamidbjss** [reply]
Awesome, thanks @benoitgrelard!

Also thanks to @mithleshjs and @gabrielmlinassi for the suggestions.

**Reply by jtabone16** [reply]
@benoitgrelard I'm still seeing this issue after updating our select component to 1.0.0. Opening the select shifts our header, which is fixed

**Comment by jtabone16** [comment]
I'm still seeing this issue after updating our select component to 1.0.0. Opening the select shifts our header, which is fixed

**Reply by benoitgrelard** [reply]
@jtabone16 I've answered in your other discussion: https://github.com/radix-ui/primitives/discussions/1586?notification_referrer_id=NT_kwDOABd_ObI0MTExNTY4NjgwOjE1Mzk4OTc&notifications_query=is%3Adone#discussioncomment-3302669

**Comment by aryanas159** [comment]
Putting modal={false} on the root dropdown element fixed it for me.

**Reply by Nikkoro** [reply]
Thank you @aryanas159 , this is the only solution that works.

**Reply by ashik-75** [reply]
but it also hide the overlay, any solution for this?

**Reply by NoCtrlZ1110** [reply]
thank you, you saved my day!

**Reply by Kochiyama** [reply]
Thank you! Life saving.

**Reply by btnalexandre** [reply]
Thankkkkks

**Reply by dpgraham4401** [reply]
Still only thing that worked for me.

**Reply by arturhaddad** [reply]
@ashik-75 to keep the overlay you can check [this solution](https://github.com/radix-ui/primitives/discussions/1100#discussioncomment-13579300)

**Comment by tspvvd** [comment]
Adding the correct `RemoveScroll` class to all relevant `position: fixed` elements fixed it for me. Good news for me I only have the navbar using `position: fixed`, but I don't think for a typical application there should be a lot of `position: fixed`. (This excludes `position: fixed` in e.g. `<DialogOverlay />`.)

```tsx
import { RemoveScroll } from 'react-remove-scroll';

<nav
  className={cn(
    'fixed inset-x-0 top-0',
    RemoveScroll.classNames.zeroRight, // here
  )}
>
  ...
</nav>
```

Ref: https://github.com/theKashey/react-remove-scroll?tab=readme-ov-file#positionfixed-elements
Ref: https://github.com/theKashey/react-remove-scroll-bar?tab=readme-ov-file#the-right-border

(This is the same thing as the answer by @/benoitgrelard above, but it wasn't immediately clear to me reading the answer at first, so I post this hopefully to make it clearer to everyone coming to this thread from Google.)

**Comment by arturhaddad** [comment]
I was able to
- keep using the `modal={true}`
- display the overlay
- and keep the scroll enabled only inside the modal

by wrapping the scrollable content within `<RemoveScroll>` using the `shards` property:

```
import * as RadixDialog from '@radix-ui/react-dialog';
import { RemoveScroll } from 'react-remove-scroll';
import { useRef } from 'react';

export default function Dialog({ open, onOpenChange, children, trigger, title }) {
  const scrollableRef = useRef(null);

  return (
    <RadixDialog.Root open={open} onOpenChange={onOpenChange} modal={true}>
      {trigger && <RadixDialog.Trigger asChild>{trigger}</RadixDialog.Trigger>}
      <RadixDialog.Portal>
        <RadixDialog.Overlay />
        <RadixDialog.Content>
          {title && <RadixDialog.Title>{title}</RadixDialog.Title>}

          {/* Using RemoveScroll with shards allows scrolling the content, with modal mode and overlay active */}
          <RemoveScroll shards={[scrollableRef]}>
            <div ref={scrollableRef}>{children}</div>
          </RemoveScroll>

        </RadixDialog.Content>
      </RadixDialog.Portal>
    </RadixDialog.Root>
  );
}

```

**Comment by kazizehsan** [comment]
I was facing this issue only on Windows, not in MacOS. I came up with the following solution.

``` 
/* always reserve space for scrollbar */
html {
  scrollbar-gutter: stable;
}
/* when radix modals are open, prevent layout shift */
body[data-scroll-locked] {
  margin-right: 0 !important;
}
```

**Reply by devzerstudiosteam** [reply]
How did you override this. since this applied from inside the radix component itself.

```
body[data-scroll-locked] {
    overflow: hidden !important;
    overscroll-behavior: contain;
    position: relative !important;
    padding-left: 0px;
    padding-top: 0px;
    padding-right: 0px;
    margin-left: 0;
    margin-top: 0;
    margin-right: 15px !important;
}
```
